integration/test/ParseObjectTest.js

```diff
@@ -1456,4 +1463,611 @@ describe('Parse Object', () => {
+        assert.equal(Object.keys(localDatastore).length, 4);
+        assert.deepEqual(localDatastore[PIN_PREFIX + 'test_pin'], [obj1._localId, obj2._localId, obj3._localId]);
+        assert.deepEqual(localDatastore[obj1._localId], obj1._toFullJSON());
+        assert.deepEqual(localDatastore[obj2._localId], obj2._toFullJSON());
+        assert.deepEqual(localDatastore[obj3._localId], obj3._toFullJSON());
+
+        await Parse.Object.saveAll(objects);
+
+        localDatastore = Parse.LocalDatastore._getLocalDatastore();
+        assert.equal(Object.keys(localDatastore).length, 4);
+        assert.deepEqual(localDatastore[PIN_PREFIX + 'test_pin'], [obj1.id, obj2.id, obj3.id]);
+        assert.deepEqual(localDatastore[obj1.id], obj1._toFullJSON());
+        assert.deepEqual(localDatastore[obj2.id], obj2._toFullJSON());
+        assert.deepEqual(localDatastore[obj3.id], obj3._toFullJSON());
+      });
+
+      it(`${controller.name} can pinAllWithName (saved)`, async () => {
+        const obj1 = new TestObject();
+        const obj2 = new TestObject();
+        const obj3 = new TestObject();
+        const objects = [obj1, obj2, obj3];
+        await Parse.Object.saveAll(objects);
+
+        Parse.Object.pinAllWithName('test_pin', objects);
+        const localDatastore = Parse.LocalDatastore._getLocalDatastore();
+        assert.equal(Object.keys(localDatastore).length, 4);
+        assert.deepEqual(localDatastore[PIN_PREFIX + 'test_pin'], [obj1.id, obj2.id, obj3.id]);
+        assert.deepEqual(localDatastore[obj1.id], obj1._toFullJSON());
+        assert.deepEqual(localDatastore[obj2.id], obj2._toFullJSON());
+        assert.deepEqual(localDatastore[obj3.id], obj3._toFullJSON());
+      });
+
+      it(`${controller.name} can unPin on destroy`, async () => {
+        const obj1 = new TestObject();
+        const obj2 = new TestObject();
+        obj1.pin();
+        obj2.pin();
+        obj1.pinWithName('test_pin');
+        obj2.pinWithName('test_pin');
+
+        await Parse.Object.saveAll([obj1, obj2]);
+
+        let localDatastore = Parse.LocalDatastore._getLocalDatastore();
+        assert(Object.keys(localDatastore).length === 4);
+        assert.equal(localDatastore[DEFAULT_PIN].includes(obj1.id), true);
+        assert.equal(localDatastore[DEFAULT_PIN].includes(obj2.id), true);
+        assert.equal(localDatastore[PIN_PREFIX + 'test_pin'].includes(obj1.id), true);
+        assert.equal(localDatastore[PIN_PREFIX + 'test_pin'].includes(obj2.id), true);
+        assert(localDatastore[obj1.id]);
+        assert(localDatastore[obj2.id]);
+
+        await obj1.destroy();
+
+        localDatastore = Parse.LocalDatastore._getLocalDatastore();
+        assert(Object.keys(localDatastore).length === 3);
+        assert.deepEqual(localDatastore[DEFAULT_PIN], [obj2.id]);
+        assert.deepEqual(localDatastore[PIN_PREFIX + 'test_pin'], [obj2.id]);
+        assert(localDatastore[obj2.id]);
```

Comment:
I would probably have used another key for the index of the objects like `${className}_${id}` 

Reply:
I like this, it will be easier to find objects